Allow completion from declared accounts

rohieb opened the pull request:
When using ledger with the `--strict` option, you need to declare all your used accounts. These declarations can additionally be used to complete account names.

rohieb commented:
Mmmh, account names can contain spaces, so I cannot safely use split('[, \t]') … will fix this tomorrow.

rohieb commented:
Should be fixed now by replacing the `split(line, '[, \t]')` with a `split(line, '\s\+;')` to remove comments at the end, and then matching to the pattern `account\s\+\(.\+\)`.

kljohann commented on autoload/ledger.vim:
```diff
@@ -267,6 +267,50 @@ endf "}}}
+  let accounts = []
+  call cursor(lnum, 0)
+  while 1
+    let lnum = search('^account\s', 'cW')
```
I think you could also use the `{stopline}` argument to `search()̀ :
```vim
while lnum != 0
    let lnum = search('^account\s', 'cW', end)
    ...
endw
```

kljohann commented on autoload/ledger.vim:
```diff
@@ -267,6 +267,50 @@ endf "}}}
+function! ledger#declared_accounts(...)
+  if a:0 == 2
+    let lnum = a:1
+    let end = a:2
```
I would have thought that `end` is a reserved keyword in vim script. But this does not seem to be the case. 😉

rohieb commented:
On 14.10.2017 18:31, Johann Klähn wrote:
> Can you re-indent your changes to match the rest of the file (inside
> while loop)?

Oh, I overlooked that.

> I think you could also use the |{stopline}| argument to `search()̀ :
> 
> while lnum != 0
>     let lnum = search('^account\s', 'cW', end)
>     ...
> endw

Yes, I think it makes sense especially with large ledger files.

> +function! ledger#declared_accounts(...)
> +  if a:0 == 2
> +    let lnum = a:1
> +    let end = a:2
> 
> I would have thought that |end| is a reserved keyword in vim script. But
> this does not seem to be the case. 😉

Mmh, vim does highlight it as a keyword when I use the variable in if
clauses... renamed to `lend` for less confusion.

Thank you for the feedback, I've updated the branch.

rohieb commented:
I'm also thinking if it would be a good idea to follow the include statements, and look for account completions in those files also. This way, when using `ledger --strict`, I could split up my journal into several files for better overview, and still make use of omnicompletion.

alerque commented:
Hey @rohieb thanks for taking the time to contribute this. I've recently stepped in to help maintenence of this plugin along a bit and am wondering what the status is. It looks like most of the PR feedback has been addressed. Is that your take too?

I agree following includes (or better yet, some way to scan all project files because in my case I'm _in the included file_ and want to complete from all accounts) but I don't think that should hold up merging this PR to add that feature (which might take a little ironing out).

Is there any reason we can't go ahead with merging this as it stands?

rohieb commented:
Thanks for coming back to this :-) I think it looks fine as-is.
